src/server/pfs/server/master.go

```diff
@@ -87,76 +92,123 @@ func (d *driver) master(ctx context.Context) {
-				if ev.Type == watch.EventError {
-					return ev.Err
+			// Watch for repo events.
+			watcher, err := postgres.NewWatcher(d.env.DB, d.env.Listener, ringPrefix, pfsdb.ReposChannelName)
+			if err != nil {
+				return errors.Wrap(err, "new watcher")
+			}
+			defer watcher.Close()
+			// Get existing entries.
+			if err := dbutil.WithTx(ctx, d.env.DB, func(ctx context.Context, tx *pachsql.Tx) error {
+				iter, err := pfsdb.ListRepo(ctx, tx, nil)
+				if err != nil {
+					return errors.Wrap(err, "create list repo iterator")
 				}
-				key := string(ev.Key)
-				lockPrefix := path.Join("repos", key)
-				if ev.Type == watch.EventDelete {
-					if cancel, ok := repos[key]; ok {
-						if err := ring.Unlock(lockPrefix); err != nil {
-							return err
-						}
-						cancel()
-						delete(repos, key)
-					}
+				return errors.Wrap(stream.ForEach[pfsdb.RepoPair](ctx, iter, func(repoPair pfsdb.RepoPair) error {
+					lockPrefix := path.Join("repos", fmt.Sprintf("%d", repoPair.ID))
+					ctx, cancel := pctx.WithCancel(ctx)
+					repos[repoPair.ID] = cancel
+					d.manageRepo(ctx, ring, repoPair.RepoInfo, lockPrefix)
 					return nil
+				}), "for each repo")
+			}, dbutil.WithReadOnly()); err != nil {
+				return errors.Wrap(err, "list repos")
+			}
+			// Process new repo events.
+			return d.handleRepoEvents(ctx, ring, repos, watcher.Watch())
+		})
+}
+
+func (d *driver) handleRepoEvents(ctx context.Context, ring *consistenthashing.Ring, repos map[pfsdb.RepoID]context.CancelFunc, watcherChan <-chan *postgres.Event) error {
+	for {
+		select {
+		case event, ok := <-watcherChan:
+			if !ok {
+				return errors.Wrap(fmt.Errorf("unexpected close on events channel"), "watch repo events")
+			}
+			if event.Err != nil {
+				log.Error(ctx, event.Err.Error())
+				continue
+			}
+			repoID := pfsdb.RepoID(event.Id)
+			lockPrefix := path.Join("repos", fmt.Sprintf("%d", event.Id))
+			if event.Type == postgres.EventDelete {
+				if cancel, ok := repos[repoID]; ok {
+					if err := ring.Unlock(lockPrefix); err != nil {
+						return err
+					}
+					cancel()
+					delete(repos, repoID)
 				}
-				if _, ok := repos[key]; ok {
+				continue
+			}
+			if _, ok := repos[repoID]; ok {
+				continue // the master has already called manageRepo for this repo.
+			}
+			ctx, cancel := pctx.WithCancel(ctx)
+			repos[repoID] = cancel
+			var repo *pfs.RepoInfo
+			var err error
+			if err := dbutil.WithTx(ctx, d.env.DB, func(ctx context.Context, tx *pachsql.Tx) error {
+				repo, err = pfsdb.GetRepo(ctx, tx, repoID)
+				return errors.Wrap(err, "get repo from event id")
+			}, dbutil.WithReadOnly()); err != nil {
+				return errors.Wrap(err, "get repo")
+			}
+			d.manageRepo(ctx, ring, repo, lockPrefix)
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (d *driver) manageRepo(ctx context.Context, ring *consistenthashing.Ring, repo *pfs.RepoInfo, lockPrefix string) {
+	key := pfsdb.RepoKey(repo.Repo)
+	go func() {
```

Comment:
Can this go statement move up to the caller.  So it's `go d.manageRepo(...)` ?  That we we don't have to indent all of this.

Reply:
sure we can do that